tests/integration/test_charm.py

```diff
@@ -0,0 +1,67 @@
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+from pathlib import Path
+
+import pytest
+import yaml
+
+
+@pytest.fixture(scope="module")
+async def charm(ops_test):
+    """Fixture for building, deploying, then removing the charm."""
+
+    built_charm = await ops_test.build_charm(".")
+    await ops_test.model.deploy(built_charm, trust=True)
+    await ops_test.model.wait_for_idle(status="active")
```

Comment:
I would lean toward this instead:

```suggestion
    await ops_test.model.wait_for_idle(raise_on_error=False)
```

That way, the fixture is agnostic to whether the charm actually succeeded or not, instead just doing the deploy and waiting for all the parts to stop moving one way or another, and the test itself can then make assertions about the "correctness".

Currently, `test_active` is actually redundant because the fixture will either timeout if the charms don't go to `active` or will raise if they go to `error` and the tests will never have a chance to inspect it. Not that big of a deal, overall, since we do generally expect the initial deploy to be ok, but as a general pattern I like the separation of concerns between fixtures providing some resource, vs tests validating something about the change.

Reply:
> the fixture is agnostic to whether the charm actually succeeded or not

I'd say that the fixture should know about whether or not the charm was successfully deployed. To quote [the pytest docs](https://docs.pytest.org/en/latest/explanation/fixtures.html#about-fixtures),

> In testing, a fixture provides a defined, reliable and consistent context for the tests. This could include environment (for example a database configured with known parameters) or content (such as a dataset).

To use that example, I would expect a `database` fixture to set up a database to the point where it's usable for testing, and not "maybe" set up a database and have the tests figure that out for themselves. Similarly here, I'd say the fixture's job is to deploy the charm, wait for it to be set up correctly, then hand it over to the tests.

Reply:
Did remove `test_active` though, you're right that it's not a useful test.